python/ray/dag/compiled_dag_node.py

```diff
@@ -1564,42 +1570,48 @@ def _execute_until(
 
         TODO(rui): catch the case that user holds onto the CompiledDAGRefs
         """
-        from ray.dag import DAGContext
+        if self._max_execution_index < execution_index:
```

Comment:
why do we need this if now?

Reply:
An ImportError will occur with the original code when destructing unused CompiledDAGRef (ones that we have not called get()). Here is a minimal repro for the original behavior:

```
foo = Foo.remote()
bar = Bar.remote()

with InputNode() as inp:
    dag = MultiOutputNode([foo.increment.bind(inp), bar.decrement.bind(inp)])

dag = dag.experimental_compile()

ref1 = dag.execute(1)
ref2 = dag.execute(1)

assert ref2.get() == [2, -2]

dag.teardown()
# When exiting the program, ref1.__del__ is invoked.
# Since it has not been called with get(), ref1.get() will
# be invoked subsequently. In _execute_until, even though
# the DAG won't be executed again, we still attempt to
# import a library, namely ray.dag.DAGContext. Since
# Python is shutting down, an ImportError occurs
# (ImportError: sys.meta_path is None, Python is likely
# shutting down).
```

The same problem persists with the current behavior if the if clause is not introduced. Another the benefit of introducing the if clause is to only import the library and calculate `timeout` when necessary. `timeout` is only used when `self._max_execution_index < execution_index` is `True`. We shouldn't import the library when `self._max_execution_index == execution_index`. I should keep the following while loop out of the if clause though.

Reply:
As I speak, I found out that if the last `CompiledDAGRef` is unused, the `ImportError` persists even when the if clause is introduced. Is there a guidance for Ray users to always invoke `get()` on all `CompiledDAGRef`? If not, this may be a bug that needs to be addressed in a separate PR. WDYT?

Here is a minimal repro:
```
foo = Foo.remote()
bar = Bar.remote()

with InputNode() as inp:
    dag = MultiOutputNode([foo.increment.bind(inp), bar.decrement.bind(inp)])

dag = dag.experimental_compile()

ref1 = dag.execute(1)
ref2 = dag.execute(1)

assert ref1.get() == [1, -1]

dag.teardown()
# Upon destruction, the DAG will be executed until the latest index. Again,
# we attempt to import DAGContext during program exit. 
```

Reply:
For now, I think it's worth keeping the if clause and have another bug tracking the dangling `CompiledDAGRef` issue.

Reply:
I think this issue I mentioned is correlated with your second point. Is there any reason why we want to avoid execution result leak (which is why we currently execute the DAG until the latest index and `get()` all results)?